### ISSUE TITLE
Support per target task parameters in plans

### ISSUE THREAD

**Issue by reidmv** [puppetlabs/bolt#1673]
## Use Case

Commonly, a large number of tasks need to be run against targets, but in such a way that a single invocation of `run_task()` is not feasible. For example, a task that requires some minor modification of parameter inputs on a target-by-target basis. Instead of one invocation of `run_task()`, a loop must be used to call `run_task()` once per node with customized inputs.

Because plans in Bolt lack any kind of parallelism features, the performance of one invocation of `run_task()` with many targets vs. many invocations of `run_task()` with one target is markedly different. A job that would take 30 seconds to run per node could draw out the time to execute a plan to 30 seconds * the number of nodes, which could easily become infeasibly long.

An example Bolt plan fragment is given below, showing what it can look like when `run_task()` needs to be called in a loop.

```puppet
['A', 'B'].each |$group| {
  getvar("compiler_${group.downcase()}_targets").each |$target| {
    run_task('peadm::agent_install', $target,
      server        => $master_target.peadm::target_name(),
      install_flags => [
        "main:certname=${target.peadm::target_name()}",
        "extension_requests:${pp_cluster}=${group}",
      ],
    )
  }
}
```

## Describe the Solution You Would Like

Bolt plans should provide some kind of parallelism feature for constructing and dispatching many tasks, but letting those tasks run in parallel and not blocking until all the dispatched tasks have completed.

Suggested implementation (per Slack conversation with @adreyer): a block-based syntax to `run_task()` which calls a block per-target to set task parameters. E.g.

```puppet
run_task(foo, $targets) |$t| {
   { param1 => 'foo', param2 => $t.name}
}
```

This provides parallelism for the use case of calling the _same_ task for a set of targets, which each need customized input parameters. While this doesn't provide unbounded parallelism power, it probably solves the 90% use case and has the strong benefit of being very simple but still decently powerful, and not dirty or ugly.

## Describe Alternatives You've Considered

Implementation approaches could include:

* passing parameters as a map to `run_task()` or `run_task_per_target()` 

      run_task(foo, $targets, _target_params => {
         "t1" => { key => val1},
         "t2" => {key => val2},
      } )

This seems more verbose and could be confusing is the keys map to groups. 

# Questions

* Should this be `run_task` or a new function `run_task_with_per_target_parameters`? Is it easy for users to understand what the block does? Is there some other case we'd want to add a block to `run_task`? As part of this ticket poll in slack to determine what the least confusing and long name for this is.
* How long is PE likely to lack support for this behavior in plans it executes? Will require DB and UI updates, and updates to the pe-bolt-server executor.
* Should it be a block that returns the parameters or a parameter that is a map of targets to parameters? Block tests better with @reidmv (add more examples for further tests). Update 04/01/2020: a block is more elegant and likely better for the common case of using facts or vars on a target.
* Should we accept defaults too? Not if it's a new function. So no.
* Do we collect all params first? Yes.
* Should we add support for other `run_*`? Not plan maybe the others. Not as part of this ticket
* Do we need to add a YAML step for this? Not now.
* What should `catch_errors` do? Test what `catch_errors` does and see if it's reasonable...
* What happens if parameters are invalid for a single target? That one target should fail, other targets should continue executing.

**Comment by nmaludy** [comment]
+1
